src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -191,13 +139,60 @@ void QgsRelationManagerDialog::addPolymorphicRelation( const QgsPolymorphicRelat
   const QList<QgsRelation> generatedRelations = polyRel.generateRelations();
   for ( const QgsRelation &generatedRelation : generatedRelations )
   {
-    if ( !generatedRelation.isValid() )
-      continue;
+    if ( addRelationPrivate( generatedRelation, item ) )
+      itemsAdded++;
+  }
+
+  mRelationsTree->setSortingEnabled( true );
+
+  return itemsAdded;
+}
+
+bool QgsRelationManagerDialog::addRelationPrivate( const QgsRelation &rel, QTreeWidgetItem *parentItem )
+{
+  if ( ! rel.isValid() )
+    return false;
+
+  QString referencingFields = rel.fieldPairs().at( 0 ).referencingField();
+  QString referencedFields = rel.fieldPairs().at( 0 ).referencedField();
+  for ( int i = 1; i < rel.fieldPairs().count(); i++ )
+  {
+    referencingFields.append( QStringLiteral( ", %1" ).arg( rel.fieldPairs().at( i ).referencingField() ) );
+    referencedFields.append( QStringLiteral( ", %1" ).arg( rel.fieldPairs().at( i ).referencedField() ) );
+  }
+
+  mRelationsTree->setSortingEnabled( false );
+  QTreeWidgetItem *item = new QTreeWidgetItem();
 
-    addRelation( generatedRelation );
+  if ( parentItem )
+  {
+    item->setFlags( item->flags() & ~Qt::ItemIsSelectable );
+  }
+  else
+  {
+    Q_ASSERT( rel.type() != QgsRelation::Generated );
```

Comment:
since we move to the type, use a switch on the type of relation, and assert that there is a parent for a generated relation?

Reply:
done